### PR TITLE
Handle truncated, corrupt, or misidentified gzip files

### PR DESCRIPTION
**Pull request by ccazabon**
Currently, when grin identifies a file as being gzipped, it tries to open it with the stdlib gzip module.  If the file is truncated or otherwise corrupt, or if it has been misidentified as containing gzipped data, then grin can die with an unhandled exception:
```
$ grin -i mail-follow-to ~/.[a-z]*
Traceback (most recent call last):
  File "~/.local/bin/grin", line 8, in <module>
    sys.exit(main())
  File "~/.local/pipx/venvs/grin3/lib/python3.9/site-packages/grin/grin.py", line 415, in main
    report = g.grep_a_file(filename, opener=openers[kind], encoding=args.encoding)
  File "~/.local/pipx/venvs/grin3/lib/python3.9/site-packages/grin/main.py", line 387, in grep_a_file
    unique_context = self.do_grep(f, encoding)
  File "~/.local/pipx/venvs/grin3/lib/python3.9/site-packages/grin/main.py", line 196, in do_grep
    block = self.read_block_with_context(None, fp, fp_size, encoding)
  File "~/.local/pipx/venvs/grin3/lib/python3.9/site-packages/grin/main.py", line 97, in read_block_with_context
    block_main = fp.read(target_io_size)
  File "/usr/lib/python3.9/gzip.py", line 300, in read
    return self._buffer.read(size)
  File "/usr/lib/python3.9/_compression.py", line 68, in readinto
    data = self.read(len(byte_view))
  File "/usr/lib/python3.9/gzip.py", line 487, in read
    if not self._read_gzip_header():
  File "/usr/lib/python3.9/gzip.py", line 435, in _read_gzip_header
    raise BadGzipFile('Not a gzipped file (%r)' % magic)
gzip.BadGzipFile: Not a gzipped file (b'\x13P')
```
These changes make grin catch this case, and retry the grep as if it was a plaintext file (which it might be, with some binary junk mixed in).  It currently emits a message to stderr when it does this, but that can be removed if you think it's noise and it should just silently do the right thing.  Output:
```
~/.cache/mozilla/firefox/zpn9cx32e.Test-Profile/cache2/entries/FE2CF3995D935F0A073049B0E2932B2D9BE8B273: Not a gzipped file (b'+\x8f'); retrying as plaintext.
```
It includes tests.

Thanks for maintaining grin3!

### CODE REVIEW

File: grin/grin.py [hunk 1]
```diff
@@ -412,7 +412,15 @@ def main(argv=None):
         g = GrepText(regex, args)
         openers = dict(text=open, gzip=gzip.open)
         for filename, kind in get_filenames(args):
-            report = g.grep_a_file(filename, opener=openers[kind], encoding=args.encoding)
+            try:
+                report = g.grep_a_file(filename, opener=openers[kind], encoding=args.encoding)
+            except (gzip.BadGzipFile, EOFError) as ex:
```
Comment (rsalmaso) [inline grin/grin.py:417]: `gzip.BadGzipFile` is only available from python 3.8, so I switch to old `OSError`

File: grin/grin.py [hunk 2]
```diff
@@ -412,7 +412,15 @@ def main(argv=None):
...
+                if kind != "gzip":
+                    # probably shouldn't happen; something weird
+                    raise
+                # Alert the user, or just silently switch to plaintext grep?
+                print(f"{filename}: {ex}; retrying as plaintext.", file=sys.stderr)
```
Comment (rsalmaso) [inline grin/grin.py:422]: `f-string` are python >= 3.6, and 2.6.X supports 3.5.
I've anyway removed the alert, so no real problem.

**Comment by rsalmaso** [comment]
Hi, thanks for this PR.
I've merged and edit a little via hg, and release a new version: https://pypi.org/project/grin3/2.6.1/

Next (feature) version will drop python 3.5, add typing support and a little refactor (but don't hold your breath :sweat_smile:)

**Comment by ccazabon** [comment]
Thanks for merging!

**Pull request closed without merging**